File: app/src/main/java/com/github/pyzahl/sofimagic/Logger.java
package com.github.pyzahl.sofimagic;

import android.os.Environment;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import android.text.format.Time;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import com.github.ma1co.openmemories.framework.DateTime;
import com.github.ma1co.pmcademo.app.BaseActivity;

import java.util.Timer;
import java.util.TimerTask;
public class Logger
{
    public static File getFile() {
        return new File(Environment.getExternalStorageDirectory(), "SOFIMAGI/LOG.TXT");
    }

    static private int level=2;

    protected static void log_time_ms(String msg) {
        try {
            String date=BaseActivity.getHMSMSfromMS(BaseActivity.getMilliSecondsOfDay());
            getFile().getParentFile().mkdirs();
            BufferedWriter writer = new BufferedWriter(new FileWriter(getFile(), true));
            writer.append(date + " " + msg);
            writer.newLine();
            writer.close();
        } catch (IOException e) {}
    }

    protected static void log(String msg) {
        try {
            Calendar calendar = DateTime.getInstance().getCurrentTime();
            String date = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(calendar.getTime());
            int ms = calendar.get(Calendar.MILLISECOND);
            getFile().getParentFile().mkdirs();
            BufferedWriter writer = new BufferedWriter(new FileWriter(getFile(), true));
            writer.append(String.format("%s.%03d %s",date,ms,msg));
            writer.newLine();
            writer.close();
        } catch (IOException e) {}
    }

    public static void set_verbose_level(int l) { level=l; };
    protected static void log(int L, String type, String msg) { if (level>L) log("[" + type + "] " + msg); }
    public static void info_debug(String msg) { log(2, "DEBUG", msg); }
    public static void info(String msg) { log(1, "INFO", msg); }
    public static void info_progress(String msg) { log(0, "INFO", msg); }
    public static void shootdata(String phase, String msg) { log(-1, "PHOTO of "+phase, msg); }
    public static void error(String msg) { log(-1, "ERROR", msg); }
}

File: gMagicPy/logger.py
import os
from datetime import datetime
from timeutil import *

class Logger:
    @staticmethod
    def get_file():
        return os.path.join(".", "LOG.TXT")

    level = 2
    system_time_offset = 0
    
    @staticmethod
    def log_time_ms(msg):
        try:
            # Assuming BaseActivity.getHMSMSfromMS(BaseActivity.getMilliSecondsOfDay()) is replaced with a Python equivalent
            date = get_hmsms_from_ms(get_milliseconds_of_day(system_time_offset))  #    datetime.now().strftime("%H:%M:%S.%f")[:-3]
            os.makedirs(os.path.dirname(Logger.get_file()), exist_ok=True)
            print(f"{date} {msg}\n")
            with open(Logger.get_file(), "a") as writer:
                writer.write(f"{date} {msg}\n")
        except IOError:
            pass

    @staticmethod
    def log(msg):
        try:
            date=get_hmsms_from_ms(get_milliseconds_of_day(system_time_offset))  # date = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
            #ms = datetime.now().microsecond // 1000
            os.makedirs(os.path.dirname(Logger.get_file()), exist_ok=True)
            #print(f"{date}.{ms:03d} {msg}\n")
            print(f"{date} {msg}\n")
            with open(Logger.get_file(), "a") as writer:
                #writer.write(f"{date}.{ms:03d} {msg}\n")
                writer.write(f"{date} {msg}\n")
        except IOError:
            pass

    @staticmethod
    def set_system_time_offset(dt):
        Logger.sysetm_time_offset = dt

    @staticmethod
    def set_verbose_level(l):
        Logger.level = l

    @staticmethod
    def get_verbose_level():
        return Logger.level

    #@staticmethod
    def log_type(L, type, msg):
        if Logger.level > L:
            Logger.log(f"[{type}] {msg}")

    @staticmethod
    def info_debug(msg):
        Logger.log_type(2, "DEBUG", msg)

    @staticmethod
    def info(msg):
        Logger.log_type(1, "INFO", msg)

    @staticmethod
    def info_progress(msg):
        Logger.log_type(0, "INFO", msg)

    @staticmethod
    def shootdata(phase, msg):
        Logger.log_type(-1, f"PHOTO of {phase}", msg)

    @staticmethod
    def error(msg):
        Logger.log_type(-1, "ERROR", msg)


